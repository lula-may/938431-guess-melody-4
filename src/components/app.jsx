import React from "react";
import {WelcomeScreen} from "./welcome-screen.jsx";

export const App = (props) => {
// eslint-disable-next-line react/prop-types
  const {errorsAmount} = props;

  return <WelcomeScreen
    errorsAmount={errorsAmount}
  />;
};
