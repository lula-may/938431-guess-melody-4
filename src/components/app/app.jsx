import React, {PureComponent} from "react";
import {BrowserRouter, Route, Switch} from "react-router-dom";
import {connect} from "react-redux";
import PropTypes from "prop-types";

import ArtistQuestionScreen from "../artist-question-screen/artist-question-screen.jsx";
import GameOverScreen from "../game-over-screen/game-over-screen.jsx";
import GameScreen from "../game-screen/game-screen.jsx";
import GenreQuestionScreen from "../genre-question-screen/genre-question-screen.jsx";
import WelcomeScreen from "../welcome-screen/welcome-screen.jsx";
import WinScreen from "../win-screen/win-screen.jsx";
import withActivePlayer from "../../hocs/with-active-player/with-active-player.jsx";
import withUserAnswer from "../../hocs/with-user-answer/with-user-answer.jsx";

import {ActionCreator} from "../../reducer.js";
import {GameType} from "../../const";

const ArtistQuestionScreenWrapped = withActivePlayer(ArtistQuestionScreen);
const GenreQuestionScreenWrapped = withActivePlayer(withUserAnswer(GenreQuestionScreen));

class App extends PureComponent {

  render() {
    const {questions} = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderGameScreen()}
          </Route>
          <Route exact path="/dev-artist">
            <ArtistQuestionScreenWrapped
              onAnswer={() => {}}
              question={questions[0]}/>
          </Route>
          <Route exact path="/dev-genre">
            <GenreQuestionScreenWrapped
              onAnswer={() => {}}
              question={questions[1]}/>
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }

  _renderGameScreen() {
    const {
      maxMistakes,
      mistakes,
      onAnswer,
      onWelcomeButtonClick,
      questions,
      resetGame,
      step,
    } = this.props;

    const question = questions[step];
    if (step === -1) {
      return (
        <WelcomeScreen
          errorsCount={maxMistakes}
          onWelcomeButtonClick={onWelcomeButtonClick}
        />
      );
    }

    if (mistakes >= maxMistakes) {
      return (
        <GameOverScreen
          onReplayButtonClick={resetGame}
        />
      );
    }

    if (step >= questions.length) {
      return (
        <WinScreen
          mistakesCount={mistakes}
          onReplayButtonClick={resetGame}
          questionsCount={questions.length}
        />
      );
    }

    if (question) {
      switch (question.type) {
        case GameType.ARTIST:
          return (
            <GameScreen
              type={question.type}
            >
              <ArtistQuestionScreenWrapped
                onAnswer={onAnswer}
                question={question}
              />
            </GameScreen>
          );

        case GameType.GENRE:
          return (
            <GameScreen
              type={question.type}
            >
              <GenreQuestionScreenWrapped
                onAnswer={onAnswer}
                question={questions[1]}
              />
            </GameScreen>
          );
      }
    }
    return null;
  }

  _handleAnswer() {
    this.setState((prevState) => ({
      step: prevState.step + 1
    }));
  }
}
const mapStateToProps = (state) => ({
  maxMistakes: state.maxMistakes,
  mistakes: state.mistakes,
  questions: state.questions,
  step: state.step,
});

const mapDispatchToProps = (dispatch) => ({
  onAnswer(question, userAnswer) {
    dispatch(ActionCreator.incrementMistake(question, userAnswer));
    dispatch(ActionCreator.incrementStep());
  },
  onWelcomeButtonClick() {
    dispatch(ActionCreator.incrementStep());
  },
  resetGame() {
    dispatch(ActionCreator.resetGame());
  }
});

App.propTypes = {
  maxMistakes: PropTypes.number.isRequired,
  mistakes: PropTypes.number.isRequired,
  onAnswer: PropTypes.func.isRequired,
  onWelcomeButtonClick: PropTypes.func.isRequired,
  questions: PropTypes.arrayOf(PropTypes.object).isRequired,
  resetGame: PropTypes.func.isRequired,
  step: PropTypes.number.isRequired,
};

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
