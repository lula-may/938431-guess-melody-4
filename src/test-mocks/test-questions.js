const questions = [
  {
    type: `singer`,
    song: {
      singer: `Lady Gaga`,
      src: `https://music/Lady_Gaga.ogg`
    },
    answers: [
      {
        singer: `Lady Gaga`,
        avatar: `https://images/Lady_Gaga.jpg`,
        id: `0`
      },
      {
        singer: `Witney Houston`,
        avatar: `https://images/Witnew_Houston.jpg`,
        id: `1`
      },
      {
        singer: `Cher`,
        avatar: `https://images/Cher.jpg`,
        id: `2`
      }
    ]
  },
  {
    type: `genre`,
    genre: `jazz`,
    answers: [
      {
        src: `https://samples/1.ogg`,
        genre: `blues`,
        id: `3`
      },
      {
        src: `https://samples/2.ogg`,
        genre: `jazz`,
        id: `4`
      },
      {
        src: `https://samples/3.ogg`,
        genre: `folk`,
        id: `5`
      },
      {
        src: `https://samples/4.ogg`,
        genre: `jazz`,
        id: `6`
      }
    ]
  }
];

export {questions};
