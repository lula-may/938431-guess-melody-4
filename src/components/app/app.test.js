import React from "react";
import rerender from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {App} from "./app";
import {questions} from "../../test-mocks/test-questions";

const mockStore = configureStore([]);

describe(`App Component`, () => {
  it(`should render WelcomeScreen`, () => {
    const store = mockStore({
      mistakes: 0
    });
    const tree = rerender.create(
        <Provider store={store}>
          <App
            maxMistakes={3}
            mistakes={0}
            questions={questions}
            onAnswer={() => {}}
            onWelcomeButtonClick={() => {}}
            resetGame={() => {}}
            step={-1}
          />
        </Provider>
    )
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render ArtistQuestionScreen`, () => {
    const store = mockStore({
      mistakes: 0
    });
    const tree = rerender.create(
        <Provider store={store}>
          <App
            maxMistakes={3}
            mistakes={2}
            questions={questions}
            onAnswer={() => {}}
            onWelcomeButtonClick={() => {}}
            resetGame={() => {}}
            step={0}
          />
        </Provider>,
        {createNodeMock: () => {
          return {};
        }}
    )
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render GenreQuestionScreen`, () => {
    const store = mockStore({
      mistakes: 0
    });
    const tree = rerender.create(
        <Provider store={store}>
          <App
            maxMistakes={3}
            mistakes={2}
            questions={questions}
            onAnswer={() => {}}
            onWelcomeButtonClick={() => {}}
            resetGame={() => {}}
            step={1}
          />
        </Provider>,
        {createNodeMock: () => {
          return {};
        }}
    )
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render GameOverScreen`, () => {
    const store = mockStore({
      mistakes: 3
    });
    const tree = rerender.create(
        <Provider store={store}>
          <App
            maxMistakes={3}
            mistakes={4}
            questions={questions}
            onAnswer={() => {}}
            onWelcomeButtonClick={() => {}}
            resetGame={() => {}}
            step={4}
          />
        </Provider>,
        {createNodeMock: () => {
          return {};
        }}
    )
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render WinScreen`, () => {
    const store = mockStore({
      mistakes: 2
    });
    const tree = rerender.create(
        <Provider store={store}>
          <App
            maxMistakes={3}
            mistakes={2}
            questions={questions}
            onAnswer={() => {}}
            onWelcomeButtonClick={() => {}}
            resetGame={() => {}}
            step={3}
          />
        </Provider>,
        {createNodeMock: () => {
          return {};
        }}
    )
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

});
