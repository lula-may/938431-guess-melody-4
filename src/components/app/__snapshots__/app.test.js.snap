// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render ArtistQuestionScreen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="answer-0"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-0"
        />
        <label
          className="artist__name"
          htmlFor="answer-0"
        >
          <img
            alt="Lady Gaga"
            className="artist__picture"
            src="https://images/Lady_Gaga.jpg"
          />
          Lady Gaga
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="answer-1"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-1"
        />
        <label
          className="artist__name"
          htmlFor="answer-1"
        >
          <img
            alt="Witney Houston"
            className="artist__picture"
            src="https://images/Witnew_Houston.jpg"
          />
          Witney Houston
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="answer-2"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-2"
        />
        <label
          className="artist__name"
          htmlFor="answer-2"
        >
          <img
            alt="Cher"
            className="artist__picture"
            src="https://images/Cher.jpg"
          />
          Cher
        </label>
      </div>
    </form>
  </section>
</section>
`;

exports[`App Component should render AuthScreen 1`] = `
<section
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="login__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="login__text"
  >
    Хотите узнать свой результат? Представтесь!
  </p>
  <form
    action=""
    className="login__form"
    onSubmit={[Function]}
  >
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="name"
      >
        Логин
      </label>
      <input
        className="login__input"
        id="name"
        name="name"
        required={true}
        type="email"
      />
    </p>
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="password"
      >
        Пароль
      </label>
      <input
        className="login__input"
        id="password"
        name="password"
        required={true}
        type="password"
      />
      <span
        className="login__error"
      >
        Неверный пароль
      </span>
    </p>
    <button
      className="login__button button"
      type="submit"
    >
      Войти
    </button>
  </form>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Сыграть ещё раз
  </button>
</section>
`;

exports[`App Component should render ErrorScreen 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
    style={
      Object {
        "color": "red",
      }
    }
  >
    Ошибка загрузки данных.
  </h2>
  <p
    className="result__total result__total--fail"
  >
    Упсс... У нас что-то сломалось. Попробуйте зайти позднее.
  </p>
</section>
`;

exports[`App Component should render GameOverScreen 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Какая жалость!
  </h2>
  <p
    className="result__total result__total--fail"
  >
    У вас закончились все попытки. Ничего, повезёт в следующий раз!
  </p>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Попробовать ещё раз
  </button>
</section>
`;

exports[`App Component should render GenreQuestionScreen 1`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите 
      jazz
       треки
    </h2>
    <form
      className="game__tracks"
      onSubmit={[Function]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="answer-3"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-3"
          />
          <label
            className="game__check"
            htmlFor="answer-3"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="answer-4"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-4"
          />
          <label
            className="game__check"
            htmlFor="answer-4"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="answer-5"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-5"
          />
          <label
            className="game__check"
            htmlFor="answer-5"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="answer-6"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-6"
          />
          <label
            className="game__check"
            htmlFor="answer-6"
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;

exports[`App Component should render WelcomeScreen 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[Function]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      Нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      3
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;

exports[`App Component should render WinScreen 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="result__total"
  >
    Вы ответили правильно на 
    0
     вопросов и совершили 
    2
     ошибки
  </p>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Сыграть ещё раз
  </button>
</section>
`;
